### ISSUE TITLE
Drop certifi, use system trust store by default

### ISSUE THREAD

**Issue by sethmlarson** [encode/httpx#302]
This is an issue that many people have been trying to tackle for Python but hasn't been done yet. I think we're in a pretty good position to try to tackle this in a way that cam be available to everyone. 

- Linux/OpenSSL should use the default system CA bundle path compiled into OpenSSL.
- Windows should use SChannel which is capable of fetching updated CA certs if they're not in the trust store.
- macOS should use SecureTransport

This should probably be implemented as a separate library, maybe use a stripped-down oscrypto project as a starting point.

Reasons to do this:
- HTTPX will use the same CA certs as the system
- Easier configuration and deployment to corporate settings
- If a system is shipped with outdated OpenSSL (Windows, macOS) we don't care because the system is more likely to be up to date than whatever was shipped with Python. 
- Windows and macOS automatically update their certs and CRLs.

**Comment by yeraydiazdiaz** [comment]
It's been a while since this was raised has been brought up recently a couple of times, could we go into detail of what's needed here?

For reference [oscrypto](https://github.com/wbond/oscrypto) lists [other related libraries](https://github.com/wbond/oscrypto#related-crypto-libraries), would any of those fit our needs? (I'd look myself but I just don't know enough about the subject 🙂 )

**Comment by florimondmanca** [comment]
Hey! So considering the amount of :+1:'s on this topic I think it would make sense to decide on whether we should keep `certifi` or use system defaults as part of 1.0 - https://github.com/encode/httpx/issues/947#issuecomment-629698226

I had a striking moment of consciousness about this earlier today:

What if we just used [`ssl.create_default_context()`](https://docs.python.org/3/library/ssl.html#ssl.create_default_context)

From the docs (emphasis mine):

> Return a new SSLContext object with default settings for the given purpose. The settings are chosen by the ssl module, and usually represent a higher security level than when calling the SSLContext constructor directly.
>
> cafile, capath, cadata represent optional CA certificates to trust for certificate verification, as in SSLContext.load_verify_locations(). If all three are None, **this function can choose to trust the system’s default CA certificates instead**.

TL;DR: it looks to me `ssl.create_default_context()` will* create an `ssl_context` that uses the system defaults. Is there anything else we need? Might draft a PR for this…

* The usage of "*can*" in the docs suggests that this might be "_may_ create", but I don't see why it wouldn't be the case. Are there cases when Python _wouldn't_ trust the default system CA?

**Comment by tomchristie** [comment]
Not my core expertise, but pretty sure it's not a solved issue.

I sat down with @lukasa last year briefly, and one of the technical issues we talked about was dropping certifi, and the complexities of accessing the system trust store on different O/S's.

I'm sure @sethmlarson has a much better handle on the current state of affairs here. I think either @tiran's or @glyph's name might also have come up in my conversation with Cory, wrt. folks having done some work in this area, but I might be getting that wrong. Kinda a PyCon type thing.

In any case, assuming I *have* got the landscape correct here, I think "use system trust store by default" sits firmly in the "make this work in an independent, tightly-scoped third party package". If and when such a package exists, then yup we can take a look at using it, otherwise it's in the realm of "this isn't a resolved issue in the Python ecosystem, certifi is the best we can do until then".

**Comment by florimondmanca** [comment]
Just found out about this discussion starting back from 2016 in the Requests repo, loads of good background there: https://github.com/psf/requests/issues/2966

**Comment by tiran** [comment]
I need to finish my prototype...

**Comment by florimondmanca** [comment]
Another data point in favor of us keeping certifi at this time: `aiohttp` not bundling certs by default seems to have been causing users a lot of pain… https://github.com/aio-libs/aiohttp/issues?q=is%3Aissue+verify+failed+

In a lot of these issues the solutions are often a mix of "disable cert validation" or "use certifi".

Eg https://github.com/aio-libs/aiohttp/issues/955

**Comment by tiran** [comment]
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=960869 should help to solve a bunch of problems. Fedora / RHEL should have root CA certificates installed. I'll verify UBI mini tomorrow.

**Comment by ofek** [comment]
@tiran How far along is that prototype? 😄

**Comment by AstraLuma** [comment]
A short-term solution might be to support pulling a CA bundle from environment variables (requests also supports this). This gives a knob for system administrators to make httpx behave.

On the prior art pile, urllib3 supports SecureTransport, but not SChannel.

Since this hasn't been mentioned in the thread, TLS has several knobs with regards to security, the big ones being algorithm selection and server certificate policies. Per the docs:

> The settings are: PROTOCOL_TLS, OP_NO_SSLv2, and OP_NO_SSLv3 with high encryption cipher suites without RC4 and without unauthenticated cipher suites. Passing SERVER_AUTH as purpose sets verify_mode to CERT_REQUIRED and either loads CA certificates (when at least one of cafile, capath or cadata is given) or uses SSLContext.load_default_certs() to load default CA certificates.

This is basically drops the horribly insecure options and presents an ok default. (I am not enough of a security expert to say how good these defaults are.) In general, OpenSSL is complex, probably more complex than necessary.

**Comment by AstraLuma** [comment]
Also, a moment of silence for PEP 543.

**Comment by ofek** [comment]
Bump.

**Comment by tomchristie** [comment]
Linking to Seth's work on this... https://github.com/sethmlarson/truststore

**Comment by dgasmith** [comment]
We're hitting this issue as well when using ZTNA, for requests were able to set the `REQUESTS_CA_BUNDLE` environment variable for requests ([source](https://github.com/psf/requests/blob/2a6f290bc09324406708a4d404a88a45d848ddf9/requests/sessions.py#L767)). Happy to make a PR here as well if an environment level is a workable solution.

We can implement at the application level; however, this will lead to inconsistent behavior as the use of HTTPX grows.

**Comment by tomchristie** [comment]
> for requests we are able to set the REQUESTS_CA_BUNDLE environment variable ([source](https://github.com/psf/requests/blob/2a6f290bc09324406708a4d404a88a45d848ddf9/requests/sessions.py#L767)). Happy to make a PR here as well if an environment level is a workable solution.

Are `SSL_CERT_FILE`/`SSL_CERT_DIR` what you're looking for?

https://www.python-httpx.org/environment_variables/#ssl_cert_file

> We're hitting this issue as well

Can you be more specific?

**Comment by tomchristie** [comment]
Using `truststore` with `httpx`...

```python
import httpx
import ssl
import truststore

ssl_context = truststore.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
client = httpx.Client(verify=ssl_context)
```

**Comment by ytjohn** [comment]
I am new to httpx and ran into this problem as well.  

> > for requests we are able to set the REQUESTS_CA_BUNDLE environment variable ([source](https://github.com/psf/requests/blob/2a6f290bc09324406708a4d404a88a45d848ddf9/requests/sessions.py#L767)). Happy to make a PR here as well if an environment level is a workable solution.
> 
> Are `SSL_CERT_FILE`/`SSL_CERT_DIR` what you're looking for?
> 
> https://www.python-httpx.org/environment_variables/#ssl_cert_file
> 
> > We're hitting this issue as well
> 
> Can you be more specific?

I'm testing out httpx and immediately ran into this issue, figured I'd add some context.  REQUESTS_CA_BUNDLE is an environment variable we can set to point to a specific bundle. In our case, we point it to the system level certificates. Since so many tools (such as pip) depend on requests under the hood, this mostly eliminates the need for relying on the certifi certificates. We bake this variable into all our images and deploy environments. If an library uses requests under the hood, we don't have rewrite it to tell it about the ca bundle location, we just have to set the REQUESTS_CA_BUNDLE variable.

My workaround is to see if REQUESTS_CA_BUNDLE is set and create an ssl context.

```
if os.getenv('REQUESTS_CA_BUNDLE'):
    ssl_context = httpx.create_ssl_context(verify=os.getenv('REQUESTS_CA_BUNDLE'))
else:
    ssl_context = httpx.create_ssl_context()

async with httpx.AsyncClient(verify=ssl_context) as client:
```

This requires setting up the ssl_context and adding verify= to the client instantiation. I would be highly in favor of httpx supporting the REQUESTS_CA_BUNDLE, or even a package specific variable (HTTPX_CA_BUNDLE). Library or tool devs won't have to implement their own workaround before distribution, and anyone wanting to opt into a non-certify provided CA bundle just has to set the appropriate environment variable.

**Comment by madkinsz** [comment]
@ytjohn as noted by Tom, `SSL_CERT_FILE` is a standard used by [OpenSSL](https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_default_verify_paths.html) and other tools like [Golang](https://pkg.go.dev/crypto/x509#SystemCertPool) — does that not do what you want?

**Comment by tomchristie** [comment]
I'd suggest that we add [this example](https://github.com/encode/httpx/issues/302#issuecomment-1368094190) to [our documentation here](https://www.python-httpx.org/third_party_packages/#gists) either verbatim or as a gist, and then close this issue.

---

*Related... I believe that our ssl context API would be better as `httpx.get(..., ssl_context=...)` than our currently `verify=.../cert=...` API... see https://github.com/encode/httpx/discussions/2521*

**Comment by AstraLuma** [comment]
Please make it clear to admins (not just developers) how to set this. The problem is that most developers only care about this in the "please forward this option" sense, and laziness/ignorance means they won't.

It is dependent on the environment the end software is used in what is the reasonable default. But as someone that's been the enterprise sysadmin that's thrice-cursed certifi's lack of overrides, the use of `truststore` would have made my job much easier.

**Comment by gpongelli** [comment]
> Using `truststore` with `httpx`...
> 
> ```python
> import httpx
> import ssl
> import truststore
> 
> ssl_context = truststore.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
> client = httpx.Client(verify=ssl_context)
> ```

you saved my day!
please add this example into documentation 👍

**Comment by wimvanleuven** [comment]
Problem with the proposed solution is that this only work for direct uses of httpx.
If using a library wrapping httpx, we have no means of injecting the truststore or SSL context, which is an added value of using env vars like REQUESTS_CA_BUNDLE, SSL_CERT_FILE or installing pip-system-certs on windows.

HTH

**Comment by mehrdad-shokri** [comment]
I'm using a package which uses httpx as it's http client. there is no constructor arguments to pass client or the httpx verify arg to it. basically stuck here.

**Comment by tomchristie** [comment]
So @sethmlarson's `truststore` looks like it's mature now.
Worth considering the switch, perhaps?

**Comment by tomchristie** [comment]
Yup, very much interested.

> Still requires Python 3.10 or later, though.

Okay, that's a consideration for us, tho we could potentially have a fallback to certifi for 3.8 and 3.9.

> Truststore has also been integrated into pip 24.2+ as the default method for verifying HTTPS certificates

I knew truststore was an option in `pip`, I didn't realise it was now on-by-default. Impressive.

**Comment by ofek** [comment]
I would love if it became the default on 3.10+ for v1.0! It's what I plan to do for Hatch anyway so it would save me time and it's what is desirable everywhere I've ever worked.

**Comment by ofek** [comment]
It's worth noting that the Rust community is also going in the direction of making system certificate verification the default with this new package that everyone is transitioning to https://github.com/rustls/rustls-platform-verifier

**Comment by zanieb** [comment]
> It's worth noting that the Rust community is also going in the direction of making system certificate verification the default...

Hm interesting. uv does not use system certificates by default (and we don't really get complaints about it).

**Comment by ofek** [comment]
It looks like you do https://github.com/astral-sh/uv/pull/1512

**Comment by omdaniel** [comment]
If httpx does not chose to use truststore by default it would be helpful to allow end-users to force the use of native certs if an environmental variable is set. This would ensure a package using httpx under the hood could avoid adding logic to their api allowing their users to override the ssl_context if a method is using httpx.

**Comment by ofek** [comment]
Relevant: https://www.digicert.com/blog/tls-certificate-lifetimes-will-officially-reduce-to-47-days

**Comment by phanduynam** [comment]
Has this issue been resolved?

**Comment by judilsteve** [comment]
Really gets my goat when package maintainers ship their own cert store.

It's my machine, I'll trust the certs *I* want to trust, not a bundle of random CAs that *you* trust. Shipping your own trust store is anti-user, and achieves *nothing* for you as the package maintainer: I have to download your trust store via an SSL connection established using *my* trust store, which evidently you *don't* trust because you're trying to override it with your own. You've created a chicken and egg problem where you have no way to verify the integrity of the certs that you want to use to verify the integrity of every subsequent request.

And then there's the endless headaches that it causes for downstream consumers running in corporate environments with MITM SSL inspection...

**Please** use the system trust store by default.

HTTPX is used internally by many other python packages: Users usually don't have direct control over the args/kwargs passed to httpx when it is used as a transitive dependency. Thankfully there are at least [some envvars](https://www.python-httpx.org/environment_variables/#ssl_cert_file) that can be set to regain control over the CAs used in this situation (as long as the package hasn't specified `trust_env=False`) and make packages that depend on HTTPX usable in corporate environments, but the system trust store should really be the default.

**Comment by glyph** [comment]
> Shipping your own trust store is anti-user, and achieves _nothing_ for you as the package maintainer

@judilsteve, package maintainers ship their own cert bundles because integrating with the OS crypto libraries properly is *really annoyingly difficult and subtle* and often provokes more problems than it solves. Everyone here wants system trust roots to be trusted by default. Yelling at maintainers to "just make good software, it does no good to make bad software" does not create progress.

Please actually *read* the comments here, and on the relevant PR https://github.com/encode/httpx/pull/3409 to understand where the holdup might be and see where you might help out.

Luckily for you, one of the big blockers according to @lovelydinosaur was that the project needed to wait for 2025-10 to drop python3.9 support first, since the relevant library (`truststore`) doesn't have support back that far, and hey look what day it is.  If you want to increase the urgency, perhaps bump that issue with information about the *relevant constraint* (i.e.: python3.9 EOL happened 4 days ago), you could ask for the patch already suggested by @cclauss to be applied. Or you could ask what you could do to move this forward.

**Comment by judilsteve** [comment]
@glyph It seems the actual change to HTTPX's code on that PR is very simple (2 lines, the rest is tests/docs/pyproject.toml/requirements.txt updates). Could the python 3.9 issue not be worked around with a conditional dependency and some graceful import handling?

```toml
# pyproject.toml
dependencies = [
    'certifi; python_version < "3.10"',
    'truststore; python_version >= "3.10"',
    ...
]
```

```python
# _config.py

def create_ssl_context(verify: ssl.SSLContext | bool = True) -> ssl.SSLContext:
    import ssl

    try:
        import truststore
    except ModuleNotFoundError:
        # Fall back to `certifi` for py3.9 and below
        import certifi
        return ssl.create_default_context(cafile=certifi.where())
    else:
        return truststore.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
```

**Comment by glyph** [comment]
> Could the python 3.9 issue not be worked around with a conditional dependency and some graceful import handling?

I'm not an httpx maintainer so I can't answer that personally. You should probably go ask on the PR. But my guess is that the workaround is to wait until 4 days ago, and drop 3.9 support. Maybe write a PR that just drops that so it can get landed as a prerequisite.
